lib/Twig/Cache/Filesystem.php

```diff
@@ -36,9 +36,8 @@ public function __construct($directory, $options = 0)
      */
     public function generateKey($name, $className)
     {
-        $hash = hash('sha256', $className);
-
-        return $this->directory.'/'.$hash[0].'/'.$hash[1].'/'.$hash.'.php';
+        // We assume the class name is suffixed by a hash (see Twig_Environment::getTemplateClass) that we can use to balance the directory tree.
+        return $this->directory.'/'.substr($className, -2).'/'.$className.'.php';
```

Comment:
using one directory instead of two saves 1 node per file. the file distribution stays the same. even though the math in #1684 is wrong, I agree that this diversity is enough.